src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,374 @@
+    auto holder = _gate.hold();
+    auto params = co_await compute_upload_parameters(range, deadline);
+    if (params.has_failure()) {
+        co_return params.as_failure();
+    }
+    _params = std::move(params.value());
+    // Create a log reader config to scan the uploaded offset
+    // range. We should skip the batch cache.
+    storage::log_reader_config reader_cfg(
+      params.value().offsets.base,
+      params.value().offsets.last,
+      priority_manager::local().archival_priority());
+    reader_cfg.skip_batch_cache = true;
+    reader_cfg.skip_readers_cache = true;
+    vlog(_ctxlog.debug, "Creating log reader, config: {}", reader_cfg);
+    auto reader = co_await _part->make_reader(reader_cfg);
+    _stream = make_reader_input_stream(
+      _ntp,
+      std::move(reader),
+      _rd_buffer_size,
+      params.value().offsets,
+      model::time_until(deadline));
+    co_return outcome::success();
+}
+
+ss::future<> segment_upload::close() {
+    if (_stream.has_value()) {
+        co_await _stream->close();
+    }
+    co_await _gate.close();
+
+    // Return units
+    _params.reset();
+}
+
+ss::future<result<upload_reconciliation_result>>
+segment_upload::compute_upload_parameters(
+  inclusive_offset_range range, model::timeout_clock::time_point deadline) {
+    auto holder = _gate.hold();
+    try {
+        auto sz = co_await _part->log()->offset_range_size(
+          range.base,
+          range.last,
+          priority_manager::local().archival_priority());
+        if (!sz.has_value()) {
+            // This means that there is not enough data in the log
+            // to satisfy the request.
+            co_return make_error_code(error_outcome::not_enough_data);
+        }
+        upload_reconciliation_result result{
+          .size_bytes = sz->on_disk_size,
+          .is_compacted = _part->log()->is_compacted(
+            range.base, sz->last_offset),
+          .offsets = inclusive_offset_range(range.base, sz->last_offset),
+        };
+        co_return result;
+    } catch (const std::invalid_argument&) {
+        // This means that the requested range.base or range.last is out of
+        // range (most likely due to truncation).
+        vlog(_ctxlog.warn, "Index out of range: {}", std::current_exception());
+        co_return make_error_code(error_outcome::out_of_range);
+
+    } catch (...) {
+        if (ssx::is_shutdown_exception(std::current_exception())) {
+            vlog(
+              _ctxlog.debug,
+              "Shutdown exception: {}",
+              std::current_exception());
+            co_return make_error_code(error_outcome::shutting_down);
+        }
+        vlog(
+          _ctxlog.error, "Unexpected exception: {}", std::current_exception());
+        co_return make_error_code(error_outcome::unexpected_failure);
+    }
+}
+
+ss::future<result<upload_reconciliation_result>>
+segment_upload::compute_upload_parameters(
```

Comment:
Nit: it looks like there are a few pairs of methods here for the two range types, is there any way to combine these? They are quite similar with only minor differences. I see:

* `archival::segment_upload::compute_upload_parameters`
* `archival::segment_upload::make_segment_upload`
* `archival::segment_upload::initialize`